homeassistant/components/mysensors.py

```diff
@@ -0,0 +1,219 @@
+    - 'path/to/.homeassistant/mysensors.json'
+    - 'path/to/.homeassistant/mysensors2.json'
+  version: '1.5'
+
+sensor:
+  platform: mysensors
+
+switch:
+  platform: mysensors
+"""
+import logging
+
+from homeassistant.helpers import (validate_config)
+
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START,
+    EVENT_HOMEASSISTANT_STOP,
+    TEMP_CELCIUS)
+
+CONF_PORT = 'port'
+CONF_DEBUG = 'debug'
+CONF_PERSISTENCE = 'persistence'
+CONF_PERSISTENCE_FILE = 'persistence_file'
+CONF_VERSION = 'version'
+
+DOMAIN = 'mysensors'
+DEPENDENCIES = []
+REQUIREMENTS = [
+    'https://github.com/theolind/pymysensors/archive/'
+    '2aa8f32908e8c5bb3e5c77c5851db778f8635792.zip#pymysensors==0.3']
+_LOGGER = logging.getLogger(__name__)
+ATTR_PORT = 'port'
+ATTR_DEVICES = 'devices'
+ATTR_NODE_ID = 'node_id'
+ATTR_CHILD_ID = 'child_id'
+ATTR_UPDATE_TYPE = 'update_type'
+
+IS_METRIC = None
+CONST = None
+GATEWAYS = None
+EVENT_MYSENSORS_NODE_UPDATE = 'MYSENSORS_NODE_UPDATE'
+
+
+def setup(hass, config):  # noqa
+    """ Setup the MySensors component. """
+    # pylint:disable=no-name-in-module
+    import mysensors.mysensors as mysensors
+
+    if not validate_config(config,
+                           {DOMAIN: [CONF_PORT]},
+                           _LOGGER):
+        return False
+
+    version = config[DOMAIN].get(CONF_VERSION, '1.4')
+
+    global CONST
+    if version == '1.4':
+        import mysensors.const_14 as const
+        CONST = const
+    elif version == '1.5':
+        import mysensors.const_15 as const
+        CONST = const
+    else:
+        import mysensors.const_14 as const
+        CONST = const
+
+    # Just assume celcius means that the user wants metric for now.
+    # It may make more sense to make this a global config option in the future.
+    global IS_METRIC
+    IS_METRIC = (hass.config.temperature_unit == TEMP_CELCIUS)
+
+    def callback_generator(port, devices):
+        """Return a new callback function. Run once per gateway setup."""
+        def node_update(update_type, nid):
+            """Callback for node updates from the MySensors gateway."""
+            _LOGGER.info('update %s: node %s', update_type, nid)
+
+            hass.bus.fire(EVENT_MYSENSORS_NODE_UPDATE, {
+                ATTR_PORT: port,
+                ATTR_DEVICES: devices,
+                ATTR_UPDATE_TYPE: update_type,
+                ATTR_NODE_ID: nid
+            })
+            return
+        return node_update
+
+    def setup_gateway(port, persistence, persistence_file):
+        """Return gateway after setup of the gateway."""
+        devices = {}    # keep track of devices added to HA
+        gateway = mysensors.SerialGateway(port,
+                                          persistence=persistence,
+                                          persistence_file=persistence_file,
+                                          protocol_version=version)
+        gateway.event_callback = callback_generator(port, devices)
```

Comment:
This was my original intention, but I got in troubles figuring out how to do that. I want to create the gateway(s) in the component and the callback function that the gateway will call. Then I need a callback function for each platform. This callback function should be called when the callback function connected to the gateway gets called. I only want to import things one way, right, from the component to the platforms, otherwise I'm thinking it'll be a dependency loop. Or is that wrong? So how do I connect the callback in the component with the callback in the platform?